MaxText/layers/decoders.py

```diff
@@ -0,0 +1,711 @@
+
+    # [batch, length, emb_dim] -> [batch, length, vocab_size]
+    if cfg.logits_via_embedding:
+      # Use the transpose of embedding matrix for logit transform.
+      logits = self.shared_embedding.attend(hidden_state)
+      if self.config.normalize_embedding_logits:
+        # Correctly normalize pre-softmax logits for this shared case.
+        logits = logits / jnp.sqrt(hidden_state.shape[-1])
+      if cfg.final_logits_soft_cap:
+        logits = logits / cfg.final_logits_soft_cap
+        logits = jnp.tanh(logits) * cfg.final_logits_soft_cap
+    else:
+      logits = linears.dense_general(
+          inputs_shape=hidden_state.shape,
+          out_features_shape=cfg.vocab_size,
+          weight_dtype=cfg.weight_dtype,
+          dtype=jnp.float32 if cfg.logits_dot_in_fp32 else cfg.dtype,  # for logit training stability
+          kernel_axes=("embed", "vocab"),
+          name="logits_dense",
+          matmul_precision=self.config.matmul_precision,
+          parameter_memory_host_offload=cfg.parameter_memory_host_offload,
+      )(
+          hidden_state
+      )  # We do not quantize the logits matmul.
+
+    if model_mode in (MODEL_MODE_PREFILL, MODEL_MODE_AUTOREGRESSIVE):
+      logits = nn.with_logical_constraint(logits, (None, None, "activation_vocab"))
+    else:
+      logits = nn.with_logical_constraint(
+          logits, ("activation_embed_and_logits_batch", "activation_length", "activation_vocab")
+      )
+
+    if self.config.cast_logits_to_fp32:
+      logits = logits.astype(jnp.float32)
+
+    return logits
+
+  @nn.compact
+  def __call__(
+      self,
+      decoder_input_tokens,
+      decoder_positions,
+      decoder_segment_ids=None,
+      deterministic=False,
+      model_mode=MODEL_MODE_TRAIN,
+      previous_chunk=None,
+      slot: Optional[int] = None,
+      page_state: Optional[page_manager.PageState] = None,
+      bidirectional_mask: Optional[Any] = None,
+      image_embeddings: Optional[jnp.ndarray] = None,
+  ):
+    cfg = self.config
+    mesh = self.mesh
+    assert decoder_input_tokens.ndim == 2  # [batch, len]
+
+    # [batch, length] -> [batch, length, emb_dim]
+    y = self._apply_embedding(decoder_input_tokens, decoder_positions, deterministic, image_embeddings, bidirectional_mask)
+
+    policy = self.get_remat_policy()
+    RemattedBlockLayers = self.set_remat_policy(self.decoder_layer, policy)
+
+    if cfg.using_pipeline_parallelism:
+      if cfg.pipeline_fsdp_ag_once:
+        partition_spec = self.pipeline_module.get_weight_sharding(
+            y, decoder_segment_ids, decoder_positions, deterministic, model_mode
+        )
+      else:
+        partition_spec = None  # This partition spec is only used for the fsdp_ag_once feature.
+      if cfg.decoder_block == DecoderBlockType.DEEPSEEK:
+        assert len(RemattedBlockLayers) == 2, "Scanned layers must have a length of 2 using deepseek."
+        dense_layer = RemattedBlockLayers[0]
```

Comment:
Are these not needed before/after this anymore?

```
layer_call_kwargs = {
    "page_state": page_state,
    "previous_chunk": previous_chunk,
    "slot": slot,
}
dense_layer = RemattedBlockLayers[0]
dense_layer.__call__ = functools.partial(dense_layer.__call__, **layer_call_kwargs)
```

Also `moe_layer.__call__ = functools.partial(moe_layer.__call__, **layer_call_kwargs)`